### PR TITLE
Kyle/unpaid

### PR DESCRIPTION
**Pull request by kyleqhua**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

# What's new in this PR
Add unpaid report calculations on the home screen and imported my old calculations over. Will edit more later

[//]: # "Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

- calculates unpaid reports: finds the amount of reports where billed > paid
- calculates to charge: finds the amount of customers that still need to be charged
- calculates outstanding payments: finds the amount of customers that still owes money.

[//]: # 'The order in which to review files and what to expect when testing locally'
home.tsx
momentUtil.js

[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"
Uses Tiffany's fin sum prs.

### Screenshots
Screen Shot 2021-03-03 at 4 17 36 PM
Screen Shot 2021-03-03 at 4 18 16 PM

[//]: # "Add screenshots of expected behavior - GIFs if you're feeling fancy!"

CC: @julianrkung

[//]: # 'This tags Julian as a default. Feel free to change, or add on anyone who you should be in on the conversation.'

### CODE REVIEW

File: src/screens/Home/Home.tsx [hunk 1]
```diff
@@ -37,6 +38,70 @@ interface HomeProps {
...
 function Home(props: HomeProps) {
   const { classes, lastUpdated, isOnline, currentSite } = props;
+  useEffect(() => {
+    proccessData();
+  }, []);
+
+  const [outStandingPayments, setOutstandingPayments] = useState<number>();
+  const [toCharge, setToCharges] = useState<number>();
+  const [customerDetails, setCustomerDetails] = useState<number>();
+  const [unpaidReports, setUnpaidReports] = useState<number>();
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: I don't think this has to be a stateful component. We can calculate the data on render and show it without using state

File: src/screens/Home/Home.tsx [hunk 2]
```diff
@@ -37,6 +38,70 @@ interface HomeProps {
...
+    setToCharges(toCharge);
+    setCustomerDetails(toCharge + outstanding);
+  }
+
+  //true has outstadining false has paid
+  const checkPayment = (customer: CustomerRecord) => {
+    if (parseInt(customer.outstandingBalance) > 0) {
+      return true;
+    }
+    return false;
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: reduce to 1 line. `return parseInt(customer.outstandingBalance) > 0`

File: src/screens/Home/Home.tsx [hunk 3]
```diff
@@ -37,6 +38,70 @@ interface HomeProps {
...
+  }
+
+  const calculateUnpaidReports = (ss: SiteRecord) => {
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: What does ss stand for? Can we just name it `site`?

File: src/screens/Home/Home.tsx [hunk 4]
```diff
@@ -37,6 +38,70 @@ interface HomeProps {
...
+  const [unpaidReports, setUnpaidReports] = useState<number>();
+
+
+
+  const proccessData = () => {
+    calculateCustomerData(currentSite);
+    calculateUnpaidReports(currentSite)
+  }
+
+  const calculateCustomerData = (ss: SiteRecord) => {
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: What does ss stand for? Can we just name it `site`?

File: src/screens/Home/Home.tsx [hunk 5]
```diff
@@ -37,6 +38,70 @@ interface HomeProps {
...
+    // customers to charge:  haven't charged yet/ no meterreading
+    // outstanding payments: done the meter reading / charged, haven't paid yet
+    let toCharge = 0;
+    let outstanding = 0;
+    //get day
+    const today = new Date();
+    const currMonth = today.getUTCMonth();
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: We should use the moment library for date handling. It'd be good to write whatever util functions you need in `momentUtils` to add a layer of abstraction. See Tiffany's PR https://github.com/calblueprint/meepanyar/pull/21 for a good example

File: src/screens/Home/Home.tsx [hunk 6]
```diff
@@ -37,6 +38,70 @@ interface HomeProps {
...
+    let customers = ss.customers;
+    for (let i = 0; i < customers.length; i++) {
+      let curr = customers[i]
+      //depends if the meter readings list is sorted with earliest => latest
+      let latestMeterReading = curr.meterReadings[curr.meterReadings.length - 1];
+      if (latestMeterReading != null) {
+        const dateTime = Date.parse(latestMeterReading.date);
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: I don't think it's sorted in redux. We shouldn't impose that invariant arbitrarily since it just gets pulled and stored directly from Airtable.

File: src/screens/Home/Home.tsx [hunk 7]
```diff
@@ -37,6 +38,70 @@ interface HomeProps {
...
+  const proccessData = () => {
+    calculateCustomerData(currentSite);
+    calculateUnpaidReports(currentSite)
+  }
+
+  const calculateCustomerData = (ss: SiteRecord) => {
+    // customers to charge:  haven't charged yet/ no meterreading
+    // outstanding payments: done the meter reading / charged, haven't paid yet
+    let toCharge = 0;
+    let outstanding = 0;
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: `outstandingPayments` is a better variable name. Outstanding is weird because it's an actual word with a main definition that isn't relevant to what it's doing

File: src/screens/Home/Home.tsx [hunk 8]
```diff
@@ -37,6 +38,70 @@ interface HomeProps {
...
+    //get day
+    const today = new Date();
+    const currMonth = today.getUTCMonth();
+    let customers = ss.customers;
+    for (let i = 0; i < customers.length; i++) {
+      let curr = customers[i]
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: Make variable `const` if it's not changing. Also choose a more descriptive variable name. `currCustomer` or something like that to show what the object is

File: src/screens/Home/Home.tsx [hunk 9]
```diff
@@ -37,6 +38,70 @@ interface HomeProps {
...
+      //depends if the meter readings list is sorted with earliest => latest
+      let latestMeterReading = curr.meterReadings[curr.meterReadings.length - 1];
+      if (latestMeterReading != null) {
+        const dateTime = Date.parse(latestMeterReading.date);
+        let readingDate = new Date(dateTime);
+        if (readingDate.getUTCMonth() < currMonth) {
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: Write a util function in the moment utils file so we have a standardized way of date comparison. The function could pass in the string dates we get from Airtable and do proper comparisons

File: src/screens/Home/Home.tsx [hunk 10]
```diff
@@ -37,6 +38,70 @@ interface HomeProps {
 
 function Home(props: HomeProps) {
   const { classes, lastUpdated, isOnline, currentSite } = props;
+  useEffect(() => {
+    proccessData();
+  }, []);
+
+  const [outStandingPayments, setOutstandingPayments] = useState<number>();
+  const [toCharge, setToCharges] = useState<number>();
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: `numOutstandingPayments` and `numCustomersToCharge` aren't too long and make this a lot easier to read imo

File: src/screens/Home/Home.tsx [hunk 11]
```diff
@@ -54,15 +119,15 @@ function Home(props: HomeProps) {
       <Link to={'/customers'}>
         <HomeMenuItem
           label="Customer Alerts"
-          amount={1}
+          amount={customerDetails ? customerDetails : 0}
           sublabels={[
             { amount: 1, label: 'Customers to Charge' },
             { amount: 2, label: 'Outstanding Payments' },
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: Change the amount in these sublabels to show the correct numbers from your calculations

**Review by julianrkung — COMMENTED** [review]
Left some comments!

File: src/screens/Home/Home.tsx [hunk 12]
```diff
@@ -38,6 +39,56 @@ interface HomeProps {
 function Home(props: HomeProps) {
   const { classes, lastUpdated, isOnline, currentSite } = props;
 
+  const calculateCustomerData = () => {
+    // customers to charge:  haven't charged yet/ no meterreading
+    // outstanding payments: done the meter reading / charged, haven't paid yet
+    let toCharge = 0;
+    let outstandingPayments = 0;
+    let customers = currentSite.customers;
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: sorry I think my feedback before wasn't the best on this. I think the clearest variable name is `numCustomersToCharge` and `numOutstandingPayments`.

`customers` should also be a `const` since it doesn't change.

File: src/screens/Home/Home.tsx [hunk 13]
```diff
@@ -38,6 +39,56 @@ interface HomeProps {
...
+          outstandingPayments += 1;
+        }
+      } else {
+        toCharge += 1;
+      }
+    }
+    console.log('payments')
+    console.log(outstandingPayments);
+    console.log('need to charge')
+    console.log(toCharge);
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: remove logs

File: src/screens/Home/Home.tsx [hunk 14]
```diff
@@ -38,6 +39,56 @@ interface HomeProps {
...
+    let total = outstandingPayments + toCharge;
+    return { 'outstandingPayments': outstandingPayments, 'toCharge': toCharge, 'totalAmount': total }
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: We should just do the addition of `totalAmount` when passing the `amount` prop to the component. `amount={customerData.oustandingPayments + customerData.toCharge}` is clearer because a `totalAmount` variable has no context/history in the app / screen other than what this function arbitrarily assigns to it which makes it unintuitive for a developer reading the code.

File: src/screens/Home/Home.tsx [hunk 15]
```diff
@@ -38,6 +39,56 @@ interface HomeProps {
...
+    let unpaid = 0;
+    let summaries = currentSite.financialSummaries;
+    for (let i = 0; i < summaries.length; i++) {
+      let singleSummary = summaries[i];
+      if (singleSummary.totalCustomersBilled > singleSummary.totalCustomersPaid) {
+        unpaid += 1;
+      }
+    }
+    console.log('unpaid')
+    console.log(unpaid)
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: remove logs

File: src/screens/Home/Home.tsx [hunk 16]
```diff
@@ -38,6 +39,56 @@ interface HomeProps {
...
+  }
+
+  //true has outstadining false has paid
+  const checkPayment = (customer: CustomerRecord) => {
+    return parseInt(customer.outstandingBalance) > 0;
+  }
+
+  const calculateUnpaidReports = () => {
+    let unpaid = 0;
+    let summaries = currentSite.financialSummaries;
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: const

File: src/screens/Home/Home.tsx [hunk 17]
```diff
@@ -38,6 +39,56 @@ interface HomeProps {
...
+    return { 'outstandingPayments': outstandingPayments, 'toCharge': toCharge, 'totalAmount': total }
+  }
+
+  //true has outstadining false has paid
+  const checkPayment = (customer: CustomerRecord) => {
+    return parseInt(customer.outstandingBalance) > 0;
+  }
+
+  const calculateUnpaidReports = () => {
+    let unpaid = 0;
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: `numUnpaidReports`

File: src/screens/Home/Home.tsx [hunk 18]
```diff
@@ -38,6 +39,56 @@ interface HomeProps {
...
+    // outstanding payments: done the meter reading / charged, haven't paid yet
+    let toCharge = 0;
+    let outstandingPayments = 0;
+    let customers = currentSite.customers;
+    for (let i = 0; i < customers.length; i++) {
+      const currCustomer = customers[i]
+      //depends if the meter readings list is sorted with earliest => latest
+      let latestMeterReading = currCustomer.meterReadings[currCustomer.meterReadings.length - 1];
+      if (latestMeterReading != null) {
+        let lessMonth = lessThanCurrentMonth(latestMeterReading.date)
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: const.

Also we should look for a better name here. `isReadingFromCurrentPeriod` or something. The `if` statement below should ideally make intuitive sense as a standalone statement for someone reading it.

File: src/lib/moment/momentUtils.ts [hunk 19]
```diff
@@ -5,3 +5,14 @@ import moment from 'moment';
 export const formatUTCDateStringToLocal = (dateStringUTC: string): string => {
   return moment(dateStringUTC).local().format('YYYY-MM-DD HH:mm:ss');
 };
+
+//takes in a string of a date time YYYY-MM-DD format. Returns true if its before the current month.
+export const lessThanCurrentMonth = (dateString: string): boolean => {
+  let lastDate = moment(dateString);
+  let lastMonth = lastDate.get('month');
+  let currentMonth = moment().get('month');
```
Comment (julianrkung) [inline src/lib/moment/momentUtils.ts]: const

File: src/lib/moment/momentUtils.ts [hunk 20]
```diff
@@ -5,3 +5,14 @@ import moment from 'moment';
...
+  if (lastMonth < currentMonth) {
+    return true;
+  }
+  return false;
```
Comment (julianrkung) [inline src/lib/moment/momentUtils.ts]: you can just do `return lastMonth < currentMonth` instead of an if else that returns true and false

File: src/screens/Home/Home.tsx [hunk 21]
```diff
@@ -38,6 +39,50 @@ interface HomeProps {
...
+  const calculateCustomerData = () => {
+    // customers to charge:  haven't charged yet/ no meterreading
+    // outstanding payments: done the meter reading / charged, haven't paid yet
+    let toCharge = 0;
+    let outstandingPayments = 0;
+    let customers = currentSite.customers;
+    for (let i = 0; i < customers.length; i++) {
+      const currCustomer = customers[i]
+      //depends if the meter readings list is sorted with earliest => latest
+      let latestMeterReading = currCustomer.meterReadings[currCustomer.meterReadings.length - 1];
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: Don't think we can depend on the meterReadings list being sorted from earliest to latest (that's not an invariant established yet in our model, but we can). We either need to enforce that "sorted" invariant from the backend or go through all the meter readings and make sure we grab the latest one here. 

Also `const` for `latestMeterReading` since it doesn't change.

File: src/screens/Home/Home.tsx [hunk 22]
```diff
@@ -38,6 +39,50 @@ interface HomeProps {
...
+      if (latestMeterReading != null) {
+        let lessMonth = lessThanCurrentMonth(latestMeterReading.date)
+        if (lessMonth) {
+          toCharge += 1;
+        }
+        if (checkPayment(currCustomer)) {
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: This 1-liner I think makes the code harder to read. I don't know what "checking" a payment means from this context (and I have to look into the function to figure it out). We should either improve the function name so that it's clear what this `if` statement actually checks, or, since the function is a 1-liner, is simple, and only used here, we could remove the function and do the logic inside of the if statement (separating it into 2 lines for clarity would help).

File: src/screens/Home/Home.tsx [hunk 23]
```diff
@@ -38,6 +39,50 @@ interface HomeProps {
...
+          outstandingPayments += 1;
+        }
+      } else {
+        toCharge += 1;
+      }
+    }
+    let total = outstandingPayments + toCharge;
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: const

File: src/screens/Home/Home.tsx [hunk 24]
```diff
@@ -38,6 +39,50 @@ interface HomeProps {
...
+    return { 'outstandingPayments': outstandingPayments, 'toCharge': toCharge, 'totalAmount': total }
+  }
+
+  //true has outstadining false has paid
+  const checkPayment = (customer: CustomerRecord) => {
+    return parseInt(customer.outstandingBalance) > 0;
+  }
+
+  const calculateUnpaidReports = () => {
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: We should ask Ashely, but I'm pretty sure unpaid reports in this context are reports that the user hasn't paid off, not that there are customers left to collect from.

The user needs to have gone through the below payment screens to have their report considered "paid" i think. 

We could possibly add a `Paid` field to see if they've paid or we could check for a copy of the bankslip. I don't know what to do for part-payments, though. I'll ask the NPO tomorrow if that's a concern.

image

File: src/lib/moment/momentUtils.ts [hunk 25]
```diff
@@ -5,3 +5,14 @@ import moment from 'moment';
 export const formatUTCDateStringToLocal = (dateStringUTC: string): string => {
   return moment(dateStringUTC).local().format('YYYY-MM-DD HH:mm:ss');
 };
+
+//takes in a string of a date time YYYY-MM-DD format. Returns true if its before the current month.
+export const lessThanCurrentMonth = (dateString: string): boolean => {
+  let lastDate = moment(dateString);
+  let lastMonth = lastDate.get('month');
```
Comment (julianrkung) [inline src/lib/moment/momentUtils.ts]: `lastMonth` and `lastDate` aren't good variable names here since they hint that it's from a month ago

File: src/lib/moment/momentUtils.ts [hunk 26]
```diff
@@ -5,3 +5,14 @@ import moment from 'moment';
 export const formatUTCDateStringToLocal = (dateStringUTC: string): string => {
   return moment(dateStringUTC).local().format('YYYY-MM-DD HH:mm:ss');
 };
+
+//takes in a string of a date time YYYY-MM-DD format. Returns true if its before the current month.
+export const lessThanCurrentMonth = (dateString: string): boolean => {
```
Comment (julianrkung) [inline src/lib/moment/momentUtils.ts]: Function doesn't consider january --> december. We need to consider periods rather than months (months aren't enough info). Tiffany merged some helpful moment functions if you want to pull (you might not even need to create a function if you're able to do the logic properly with Tiffany's existing util functions).

Comment (kyleqhua) [reply src/lib/moment/momentUtils.ts]: oh I didnt thake this into consideration ty

File: src/lib/airtable/schema.js [hunk 27]
```diff
@@ -1,6 +1,6 @@
 /*
-    THIS IS A GENERATED FILE
-    Changes might be overwritten in the future, edit with caution!
+	THIS IS A GENERATED FILE
```
Comment (julianrkung) [inline src/lib/airtable/schema.js]: Can you revert the formatting changes made here? We don't want to change the generated file apart from apply patches, or else we risk patch conflicts (because they depend on diffs).

File: src/lib/moment/momentUtils.ts [hunk 28]
```diff
@@ -6,6 +6,18 @@ export const formatUTCDateStringToLocal = (dateStringUTC: string): string => {
   return moment(dateStringUTC).local().format('YYYY-MM-DD HH:mm:ss');
 };
 
+
+export const lessThanCurrentPeriod = (dateString: string): boolean => {
+  const currentDate = moment(dateString);
+  const currentMonth = moment().startOf('month');
+  const currentDatePeriod = currentDate.valueOf();
+  const currentPeriod = currentMonth.valueOf();
```
Comment (julianrkung) [inline src/lib/moment/momentUtils.ts]: Naming is confusing here. It makes it especially hard to follow because `currentMonth` refers to the actual current month, but `currentDate` refers to the date given after parsing the date string (which could be past dates). 

We need a better name than `currentDate` and etc so that the logic is easy to read and understand

File: src/lib/moment/momentUtils.ts [hunk 29]
```diff
@@ -6,6 +6,18 @@ export const formatUTCDateStringToLocal = (dateStringUTC: string): string => {
...
+  if (currentDatePeriod < currentPeriod) {
+    return true;
+  }
+  return false;
```
Comment (julianrkung) [inline src/lib/moment/momentUtils.ts]: instead of `if (...) return true else false` just return the evaluation of the if statement

File: src/screens/Home/Home.tsx [hunk 30]
```diff
@@ -1,11 +1,12 @@
-import React from 'react';
+import React, { useEffect, useState } from 'react';
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: Remove unused imports

File: src/screens/Home/Home.tsx [hunk 31]
```diff
@@ -38,6 +39,45 @@ interface HomeProps {
...
+    let numCustomersToCharge = 0;
+    let numOutstandingPayments = 0;
+    const customers = currentSite.customers;
+    for (let i = 0; i < customers.length; i++) {
+      const currCustomer = customers[i]
+      //depends if the meter readings list is sorted with earliest => latest
+      let latestMeterReading = currCustomer.meterReadings[currCustomer.meterReadings.length - 1];
+      if (latestMeterReading != null) {
+        let lessMonth = lessThanCurrentPeriod(latestMeterReading.date);
+        if (lessMonth) {
```
Comment (julianrkung) [inline src/screens/Home/Home.tsx]: https://github.com/calblueprint/meepanyar/pull/36#discussion_r586927473

You can put the function call inside the `if` statement since the variable is only used once if thinking of a good name is difficult here

**Review by julianrkung — COMMENTED** [review]
More comments! Good improvement :) .

**Review by julianrkung — COMMENTED** [review]
I think we can merge after the unresolved comments are taken care of

File: src/lib/airtable/airtable.js [hunk 32]
```diff
@@ -177,7 +177,7 @@ function getRecordsByAttribute(
       view: VIEW,
       filterByFormula: filterByFormula
         ? `AND(${filterByFormula}, {${fieldType}}=${fieldValue})`
-        : `{${fieldType}}=${fieldValue}`,
+        : `({${fieldType}}="${fieldValue}")`,
```
Comment (julianrkung) [inline src/lib/airtable/airtable.js]: Can this be deleted? Was this a mistake

**Review by julianrkung — APPROVED** [review]
Looks good! :) We can merge and work on revamping the home page now
